app/src/main/java/com/demo/recyclerviewdemo/ContactsAdapter.java

```diff
@@ -13,23 +15,31 @@
 
 import java.util.List;
 
-public class ContactsAdapter extends RecyclerView.Adapter<ContactsAdapter.ViewHolder> {
+public class ContactsAdapter extends ListAdapter<Contact, ContactsAdapter.ViewHolder> {
 
     private List<Contact> mContacts = new ArrayList<>();
 
+    public ContactsAdapter() {
+        super(DIFF_CALLBACK);
+    }
+
     public void addMoreContacts(List<Contact> newContacts) {
-        int insertionPosition = mContacts.size();
         mContacts.addAll(newContacts);
-
-        // OMIT to illustrate how the UI is independent from the dataset
-        // and does not update automatically.
-        notifyItemRangeInserted(insertionPosition, newContacts.size());
+        submitList(mContacts); // DiffUtil takes care of the chekc
     }
 
-    @Override
-    public int getItemCount() {
-        return mContacts.size();
-    }
+    public static final DiffUtil.ItemCallback<Contact> DIFF_CALLBACK =
```

Comment:
Let's define this above the function definitions, where other member fields are defined.. ?